In [1]:
import pandas as pd
path = "../Atividades/Atividade 5/"

In [2]:
df = pd.read_excel(path+"Online Retail.xlsx")
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

<H4> 1. Quantos objetos do conjunto de dados possuem o valor do atributo
InvoiceNo ausente ou nulo?

In [5]:
pd.isnull(df["InvoiceNo"]).value_counts()

False    541909
Name: InvoiceNo, dtype: int64

<p> Neste caso, vemos que nenhum destes valores é nulo

<h4> 2. Quantos objetos do conjunto de dados possuem o valor do atributo
UnitPrice menor que zero? Apague esses objetos do conjunto de dados.
    

In [7]:
df[df['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299983   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

<p> Estes são os valores que unitPrice são menores que 0

In [14]:
print("Tamanho do df: "+str(len(df)))

print("Tamanho do df sem os valores menores que 0: "+str(len(df[df['UnitPrice'] >= 0])))
df = df[df['UnitPrice'] >= 0]
print("Tamanho do df final: "+str(len(df)))


Tamanho do df: 541909
Tamanho do df sem os valores menores que 0: 541907
Tamanho do df final: 541907


<h4> 3. Qual o cliente (atributo CustomerId) que comprou mais itens, sendo que
cada linha representa um item comprado?

In [20]:
df["CustomerID"].value_counts().head(1)

17841.0    7983
Name: CustomerID, dtype: int64

<p> No caso, seria o cliente com o ID 17841.0

<h4> 4. Qual o produto mais caro vendido?

<p> O produto mais caro: 

In [23]:
df["UnitPrice"].max()

38970.0

<p> O produto mais barato:

In [24]:
df["UnitPrice"].min()

0.0

<h4> 5. Quais os países (Country) que fazem parte deste conjunto de dados?

In [15]:
df["Country"].value_counts()

United Kingdom          495476
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58
Lebanon 

<h4> 6. Qual a compra (InvoiceNo) que possui mais itens?


In [26]:
df["InvoiceNo"].value_counts().head(1)

573585    1114
Name: InvoiceNo, dtype: int64

<p> No caso, seria o InvoiceNo: 573585    

<h4> 7. Qual a compra (InvoiceNo) que tem o maior valor total?

<p> Primeiro, vamos criar uma coluna com o valor total

In [29]:
df["ValTotal"] = df["UnitPrice"] * df['Quantity']
df.head()

C:\Users\raito\AppData\Local\Temp/ipykernel_16408/802362817.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["ValTotal"] = df["UnitPrice"] * df['Quantity']


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  ValTotal  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom     15.30  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom     20.34  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom     22.00  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom     20.34  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom     20.34

In [34]:
df.sort_values("ValTotal", ascending=False).head(1)

InvoiceNo StockCode                  Description  Quantity  \
540421    581483     23843  PAPER CRAFT , LITTLE BIRDIE     80995   

               InvoiceDate  UnitPrice  CustomerID         Country  ValTotal  
540421 2011-12-09 09:15:00       2.08     16446.0  United Kingdom  168469.6

<p> Ai está a compra com o maior valor, com InvoiceNo = 581483